handbook/python-dataplugins/python-dataplugins.md

```diff
@@ -0,0 +1,264 @@
+   """
+      Returns a value array of the correct data type specified in the tdm dictionary
+   """
+```
+
+<details>
+<summary>Example Code</summary>
+<p>
+
+```python
+def read_channel_values(self, grp_index, chn_index, numberToSkip, numberToTake):
+   dataType = self.tdm_tree["groups"][grp_index]["channels"][chn_index]["type"]
+   values = []
+   for row in self.data:
+      value = row[self.channelNames[chn_index]]
+      values.append(value)
+   return values[numberToSkip:numberToTake+numberToSkip]
+```
+
+</p>
+</details>
+
+The client applications are calling that function to retrieve channel values (or a subset of values). The DataPlugin needs to implement the function to return the values for a given group and channel index. It also needs to ensure only the correct subset of values is returned for a given `numberToSkip` and `numberToTake`.
+
+Additionally, a callback function to retrieve the channel length must be implemented. In the simple case where all channels have the same length, it can simply return a constant.
+
+```python 
+def read_channel_length(self, grp_index, chn_index):
+   """
+      Returns the channel length as an Integer for a given group and channel index
+   """
+   return 10
+```
+
+See full example: [csv-read-with-callback-loading](../../examples/python-dataplugin-examples/csv-read-with-callback-loading/Readme)
+
+## Error Handling
+Python DataPlugins can raise errors in all callback functions. The raised error will be transported to DIAdem, LabVIEW or SystemLink DataFinder.
+
+```python
+def read_store(self, parameter):
+   ...
+   raise Exception("Leave read_store with exception")
+```
+
+### Not My File
+A special case is "Not My File". This error should be raised when the DataPlugin detects that the file to be opened, is not suited for this DataPlugin. This special error can be raised only in the `read_store` function by returning `None`:
+
+```python
+def read_store(self, parameter):
+   ...
+   if notMyFile:
+      return None
+```
+
+## Export DataPlugin
+Export Python DataPlugins to make them available on other systems. Use DIAdem to export a DataPlugin as a URI file.
+
+<img alt="Exporting DataPlugins in DIAdem" src="../img/pydp-diadem_export.png?raw=true" width="600"><br>
+
+## Known Limitations
+### Numpy and Pandas
+Unfortunately, Numpy and Pandas are not well supported to run in embedded Python environments and, therefore, cannot be used in DataPlugins.
+
+### Single file DataPlugins
+Python DataPlugins can only be written in a single python file. Importing sidecar files is not supported. It will fail when exporting the DataPlugin as a URI.
+
+### datetime.strptime
+There is an <a href="https://bugs.python.org/issue27400">open issue</a> in Python for `datetime.strptime` that prevents the function to work properly in embedded Python environments. Therefore, this function should be avoided in DataPlugin source code. Instead, the following function can be added to the code and work around the issue:

```

Comment:
> Therefore, this function should be avoided in DataPlugin source code. Instead, the following function can be added to the code and work around the issue:

Avoid using this function in DataPlugin source code. Instead, add the following function to your code:

Reply:
thanks